clippy_lints/src/methods/needless_character_iteration.rs

```diff
@@ -55,16 +59,27 @@ fn handle_expr(
                 && let Some(last_chain_binding_id) =
                     get_last_chain_binding_hir_id(first_param, block.stmts)
             {
-                handle_expr(cx, block_expr, last_chain_binding_id, span, before_chars, revert);
+                handle_expr(
+                    cx,
+                    block_expr,
+                    last_chain_binding_id,
+                    span,
+                    before_chars,
+                    revert,
+                    is_all,
+                );
             }
         },
-        ExprKind::Unary(UnOp::Not, expr) => handle_expr(cx, expr, first_param, span, before_chars, !revert),
+        ExprKind::Unary(UnOp::Not, expr) => handle_expr(cx, expr, first_param, span, before_chars, !revert, is_all),
         ExprKind::Call(fn_path, [arg]) => {
             if let ExprKind::Path(path) = fn_path.kind
                 && let Some(fn_def_id) = cx.qpath_res(&path, fn_path.hir_id).opt_def_id()
                 && match_def_path(cx, fn_def_id, &["core", "char", "methods", "<impl char>", "is_ascii"])
                 && path_to_local_id(peels_expr_ref(arg), first_param)
                 && let Some(snippet) = snippet_opt(cx, before_chars)
+                // If we have `!is_ascii`, then only `.any()` should warn. And if the condition is
+                // `is_ascii`, then only `.all()` should warn.
+                && revert != is_all
```

Comment:
Could you lift this to avoid running the expensive conditions unnecessarily? (Same with the other condition)

Reply:
Good point!